### PR TITLE
WriteAtPrepare: Efficient read from snapshot list

### PR DESCRIPTION
**Pull request by maysamyabandeh**
Divide the old snapshots to two lists: a few that fit into a cached array and the rest in a vector, which is expected to be empty in normal cases. The former is to optimize concurrent reads from snapshots without requiring locks. It is done by an array of std::atomic, from which std::memory_order_acquire reads are compiled to simple read instructions in most of the x86_64 architectures.

### CODE REVIEW

File: utilities/transactions/pessimistic_transaction_db.cc [hunk 1]
```diff
@@ -600,7 +601,39 @@ void WritePreparedTxnDB::AddCommitted(uint64_t prepare_seq,
       {
         WriteLock wl(&snapshots_mutex_);
         InstrumentedMutex(db_impl_->mutex());
-        snapshots_ = db_impl_->snapshots().GetAll();
+        // We only care about snapshots lower then max
+        auto all_snapshots =
```
Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: `all_snapshot` is not used if `curr_seq == snapshots_version_`. Move it into the if-block.
Also I think you only need to hold db_impl_->mutex() for snapshots()->GetAll(), but not the rest of logic?

File: utilities/transactions/pessimistic_transaction_db.cc [hunk 2]
```diff
@@ -600,7 +601,39 @@ void WritePreparedTxnDB::AddCommitted(uint64_t prepare_seq,
...
+          // we simply insert the new snapshots in order, if an overwritten item
+          // is still valid in the new list is either equal to the new one or
+          // already inserted before. This guarantess that a reader will who
+          // sees the list in the middle of update, will always see all the old
+          // snaphots that are still valid.
+          size_t i = 0;
+          auto it = all_snapshots.begin();
+          for (; it != all_snapshots.end() && i < SNAPSHOT_CACHE_SIZE;
+               it++, i++) {
+            snapshot_cache_[i].store(*it, std::memory_order_release);
```
Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: how can `snapshot_cache_` be safe for concurrent read in the middle of the loop here?

Comment (maysamyabandeh) [inline utilities/transactions/pessimistic_transaction_db.cc]: What do you mean? It is a std::atomic.

File: utilities/transactions/pessimistic_transaction_db.cc [hunk 3]
```diff
@@ -681,5 +720,20 @@ bool WritePreparedTxnDB::ExchangeCommitEntry(uint64_t indexed_seq,
   return true;
 }
 
+bool WritePreparedTxnDB::MaybeUpdateOldCommitMap(const uint64_t& prep_seq,
+                                                 const uint64_t& commit_seq,
+                                                 const uint64_t& snapshot_seq) {
+  if (commit_seq <= snapshot_seq) {
```
Comment (maysamyabandeh) [inline utilities/transactions/pessimistic_transaction_db.cc]: If we do not store an entry in old_commit_map we assume it is committed in all snapshots. if commit_seq == snapshot_seq, it is considered already in the snapshot so we need not to keep the entry around.

File: utilities/transactions/pessimistic_transaction_db.cc [hunk 4]
```diff
@@ -681,5 +720,20 @@ bool WritePreparedTxnDB::ExchangeCommitEntry(uint64_t indexed_seq,
...
+    return false;
+  }
+  // then snapshot_seq < evicted.commit_seq
+  if (prep_seq <= snapshot_seq) {  // overlapping range
+    WriteLock wl(&old_commit_map_mutex_);
+    old_commit_map_empty_.store(false, std::memory_order_release);
+    old_commit_map_[prep_seq] = commit_seq;
+  }
```
Comment (maysamyabandeh) [inline utilities/transactions/pessimistic_transaction_db.cc]: true

File: utilities/transactions/pessimistic_transaction_db.cc [hunk 5]
```diff
@@ -600,7 +601,39 @@ void WritePreparedTxnDB::AddCommitted(uint64_t prepare_seq,
       {
         WriteLock wl(&snapshots_mutex_);
         InstrumentedMutex(db_impl_->mutex());
-        snapshots_ = db_impl_->snapshots().GetAll();
+        // We only care about snapshots lower then max
```
Comment (maysamyabandeh) [inline utilities/transactions/pessimistic_transaction_db.cc]: the code is going to change a lot during the next few weeks. i am planning a refactoring phase once it is more stable.

File: utilities/transactions/pessimistic_transaction_db.cc (hunk 3 above)
Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: I see.

File: utilities/transactions/pessimistic_transaction_db.cc (hunk 5 above)
Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: ok

File: utilities/transactions/pessimistic_transaction_db.cc (hunk 2 above)
Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: Ah, sorry, what I was asking is answered in the line 612-619 comment.

File: utilities/transactions/pessimistic_transaction_db.cc [hunk 6]
```diff
@@ -600,7 +601,39 @@ void WritePreparedTxnDB::AddCommitted(uint64_t prepare_seq,
...
+        auto all_snapshots =
+            db_impl_->snapshots().GetAll(nullptr, max_evicted_seq);
+        auto curr_seq = db_impl_->GetLatestSequenceNumber();
+        // This is to avoid updating the snapshots_ if it already updated with a
+        // more recent vesion by a concrrent thread
+        if (curr_seq > snapshots_version_) {
+          snapshots_version_ = curr_seq;
+          // Both new and old lists are sorted and the new list is subset of the
+          // previous list plus some new items. Thus if a snapshot repeats in
+          // both new and old lists, it will appear upper in the new list. So if
```
Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: Why it will appear upper in the new list? Snapshots can be released out of order, say you have
old list: 1,2,3
and then 1 is released then you end up with 
new list: 2,3

File: utilities/transactions/pessimistic_transaction_db.cc (hunk 2 above)
Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: btw. Do you plan to have each read hold a snapshot explicitly? AFAIW right now it doesn't hold snapshot (i.e. won't add a snapshot to db_impl->snapshots()).

File: utilities/transactions/pessimistic_transaction_db.cc [hunk 7]
```diff
@@ -600,7 +601,39 @@ void WritePreparedTxnDB::AddCommitted(uint64_t prepare_seq,
...
+          // we simply insert the new snapshots in order, if an overwritten item
+          // is still valid in the new list is either equal to the new one or
+          // already inserted before. This guarantess that a reader will who
+          // sees the list in the middle of update, will always see all the old
+          // snaphots that are still valid.
```
Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: Say you have
```
old list: 10, 20, 30
```
and you have a reader iterating the list and reading snapshot 20. And you update the list to:
```
new list: 30, 40, 50
```
the reader resumes after the list is fully updated. It then reads 50 and miss 30 and 40.

File: utilities/transactions/pessimistic_transaction_db.cc (hunk 2 above)
Comment (maysamyabandeh) [inline utilities/transactions/pessimistic_transaction_db.cc]: When a transaction starts the API allows the users to take snapshots explicitly. If not, it takes a snapshot the first time they do a read. So all read transaction will take at least one snapshot.

File: utilities/transactions/pessimistic_transaction_db.cc (hunk 7 above)
Comment (maysamyabandeh) [inline utilities/transactions/pessimistic_transaction_db.cc]: 50 is ok to be missed since it was not an old snapshot when this thread started. In other words, 50 is bigger than the max of this thread so the evicted entry is definitely in the snapshot of 50.
But you are right about 30. The problem will be fixed if the reader reads the list twice. I will update the code accordingly.

File: utilities/transactions/pessimistic_transaction_db.cc (hunk 1 above)
Comment (maysamyabandeh) [inline utilities/transactions/pessimistic_transaction_db.cc]: it is easier to reason about the freshness of snapshot list if it is obtained atomically with the sequence number. I will add comment to the code.

File: utilities/transactions/pessimistic_transaction_db.cc (hunk 2 above)
Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: How about a plain Get()? I don't see it create a snapshot right now: https://github.com/facebook/rocksdb/blob/3c327ac2d0fd50bbd82fe1f1af5de909dad769e6/utilities/write_batch_with_index/write_batch_with_index.cc#L750

Comment (maysamyabandeh) [inline utilities/transactions/pessimistic_transaction_db.cc]: Did you take into account SetSnapshotIfNeeded() in TryLock?

Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: Where do we TryLock for Get()?

Comment (maysamyabandeh) [inline utilities/transactions/pessimistic_transaction_db.cc]: you are right. I mistook your comment with WriteBatch.

So going back to your question, no if the user does not want to take a snapshot we do not do it. What do you think we should do that?

Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: I might be paranoid, but in the read path shall we check if snapshot_seq is <= max_evited_seq_ and not among any explicit snapshots, we should let it try again?

Comment (maysamyabandeh) [inline utilities/transactions/pessimistic_transaction_db.cc]: We need to be more clear. To which snapshot_seq you are referring to? Which line in the read path you are referring to?

File: utilities/transactions/pessimistic_transaction_db.cc [hunk 8]
```diff
@@ -581,16 +581,23 @@ bool WritePreparedTxnDB::IsInSnapshot(uint64_t prep_seq,
   return false;
 }
 
-void WritePreparedTxnDB::AddPrepared(uint64_t seq) { prepared_txns_.push(seq); }
+void WritePreparedTxnDB::AddPrepared(uint64_t seq) {
+  WriteLock wl(&prepared_mutex_);
+  prepared_txns_.push(seq);
+  ROCKS_LOG_INFO(db_impl_->GetDBOptions().info_log, "Txn %lu Prepared", seq);
```
Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: better cache info_log. DBImpl::GetDBOptions() return a copy of the options, not a reference.

File: utilities/transactions/pessimistic_transaction_db.cc (hunk 2 above)
Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: Let me rephrase. We are storing list of snapshots and evict items from old_commit_map based on it. But it doesn't handle plain Get() which don't hold snapshots. But yes we haven't implement the whole read path yet other than IsInSnpashot(). We'll worry about it later.

File: utilities/transactions/pessimistic_transaction_db.cc (hunk 8 above)
Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: s/%lu/%PRIu64/

File: utilities/transactions/pessimistic_transaction_db.cc (hunk 2 above)
Comment (maysamyabandeh) [inline utilities/transactions/pessimistic_transaction_db.cc]: Our implementation needs to be consistent with the existing one. Currently if txn does not take snapshot before Get (which would be odd) then there is no guaranteed that it reads the same value in the next Get. Not storing items in old_commit_map essentially means that are not guaranteeing that either.

Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: Let's postpone the discussion to when the full read path is implemented.

File: utilities/transactions/pessimistic_transaction_db.cc (hunk 8 above)
Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: Should this be just DEBUG level?

File: utilities/transactions/pessimistic_transaction_db.cc [hunk 9]
```diff
@@ -621,18 +674,34 @@ void WritePreparedTxnDB::AddCommitted(uint64_t prepare_seq,
...
+    // The list might get updated concurrently as we are reading from it. The
+    // reader should be able to read all the snapshots before the update by the
+    // writer that are still valid. As explained above if a reader jumps in the
+    // middle of update it still sees all the old snapshots that are still
+    // valid. If a writer jumps in the middle of a read, however, the reader has
+    // to re-read all updated items by writer since the snapshots could move up
+    // in the new sorted list. To cover this case, the readers reads the list
+    // twice.
+    size_t i = 0;
+    for (int pass = 0; pass < 2; pass++) {
```
Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: I don't think scanning the list N-times solve the problem. There can be a second writer jumps in the middle on the second pass, and so on so forth.

File: utilities/transactions/pessimistic_transaction_db.h [hunk 10]
```diff
@@ -114,8 +114,10 @@ class PessimisticTransactionDB : public TransactionDB {
       Transaction* txn, const WriteOptions& write_options,
       const TransactionOptions& txn_options = TransactionOptions());
   DBImpl* db_impl_;
+  std::shared_ptr<Logger> info_log;
```
Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.h]: nit: subfix with underscore?

File: utilities/transactions/pessimistic_transaction_db.cc (hunk 9 above)
Comment (maysamyabandeh) [inline utilities/transactions/pessimistic_transaction_db.cc]: You are right. It will be fixed if the reader reads backward from the end to the front. Since the survived snapshots always move up in the list, a reader that reads from the bottom to the top of the list will eventually read that.

Comment (maysamyabandeh) [inline utilities/transactions/pessimistic_transaction_db.cc]: You are right. It would be fixed if the reader reads the list from bottom up. Since a survived snapshot always moves up in the list and a writer writes top to bottom, a survived entry is written on the higher place before it gets overwritten. So a reader who reads from the bottom will see the survived entry that is moving up at some point.

Comment (yiwu-arbug) [inline utilities/transactions/pessimistic_transaction_db.cc]: It sounds it will work with the atomic int list `snapshot_cache_` alone. But how if a snapshot moves up from `snapshots_` to `snapshot_cache_` in the middle of the scan?

I think the split of the list into `snapshot_cache_` and `snapshots_` make it too tricky now. How about we postpone the change? Since there are no concurrent commits in the MyRocks case, the `snapshots_mutex_` might not be a bottleneck.

Comment (maysamyabandeh) [reply utilities/transactions/pessimistic_transaction_db.cc]: You are right again. The fix however seems to be trivial: re-read the `snapshot_cache_` while holding the `snapshots_mutex_` for reading the `snapshots_`.

**Comment by yiwu-arbug** [comment]
Travis failure related to blob db will be fixing by #2787
